### PR TITLE
Fix SearchableSnapshotIndexMetadataUpgrader

### LINKED ISSUE #71973: [CI] SearchableSnapshotIndexMetadataUpgraderTests.testUpgradeIndices
**Build scan**:
https://gradle-enterprise.elastic.co/s/hfll4qbiunmmy (7.x)
https://gradle-enterprise.elastic.co/s/7kqvtj64bdnew (master)
**Repro line**:

(from 7.x)
```
./gradlew ':x-pack:plugin:searchable-snapshots:test' --tests "org.elasticsearch.xpack.searchablesnapshots.upgrade.SearchableSnapshotIndexMetadataUpgraderTests.testUpgradeIndices" -Dtests.seed=26B49F246A6B424 -Dtests.locale=sr-Latn -Dtests.timezone=Africa/Bissau -Druntime.java=8
```

**Reproduces locally?**:
yes

**Applicable branches**:
7.x, master

**Failure history**:
Only failed [twice so far](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-7d,mode:quick,to:now))&_a=(columns:!(_source),index:b646ed00-7efc-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:testUpgradeIndices),sort:!(process.time-start,desc))) but was recently added.

**Failure excerpt**:

org.elasticsearch.xpack.searchablesnapshots.upgrade.SearchableSnapshotIndexMetadataUpgraderTests > testUpgradeIndices FAILED
    java.lang.AssertionError: 
    Expected: "frozen"
         but: was "normal"
        at __randomizedtesting.SeedInfo.seed([26B49F246A6B424:D64E14474C5532D1]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.elasticsearch.xpack.searchablesnapshots.upgrade.SearchableSnapshotIndexMetadataUpgraderTests.lambda$testUpgradeIndices$0(SearchableSnapshotIndexMetadataUpgraderTests.java:70)
        at java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
        at java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1812)
        at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
        at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
        at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.anyMatch(ReferencePipeline.java:449)
        at org.elasticsearch.xpack.searchablesnapshots.upgrade.SearchableSnapshotIndexMetadataUpgraderTests.testUpgradeIndices(SearchableSnapshotIndexMetadataUpgraderTests.java:58)

### PR DESCRIPTION
**Pull request by henningandersen**
The upgrader that applies the shard limit group had a flaw in that it
might not upgrade indices that were mounted during rolling upgrades.

Closes #71973

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java [hunk 1]
```diff
@@ -92,7 +92,7 @@ public void onFailure(String source, Exception e) {
 
     static boolean needsUpgrade(ClusterState state) {
         return StreamSupport.stream(state.metadata().spliterator(), false)
-            .filter(imd -> imd.getCreationVersion().onOrAfter(Version.V_7_12_0) && imd.getCreationVersion().before(Version.V_7_13_0))
+            .filter(imd -> imd.getCreationVersion().onOrAfter(Version.V_7_12_0) && imd.getCreationVersion().before(Version.V_8_0_0))
```
Comment (henningandersen) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java:95]: The 7.x backport will remove this upper limit.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java [hunk 2]
```diff
@@ -104,7 +104,7 @@ static ClusterState upgradeIndices(ClusterState currentState) {
         }
         Metadata.Builder builder = Metadata.builder(currentState.metadata());
         StreamSupport.stream(currentState.metadata().spliterator(), false)
-            .filter(imd -> imd.getCreationVersion().onOrAfter(Version.V_7_12_0) && imd.getCreationVersion().before(Version.V_7_13_0))
+            .filter(imd -> imd.getCreationVersion().onOrAfter(Version.V_7_12_0) && imd.getCreationVersion().before(Version.V_8_0_0))
```
Comment (henningandersen) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java:107]: The 7.x backport will remove the upper limit (v8).

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgraderTests.java [hunk 3]
```diff
@@ -88,20 +111,31 @@ private Settings normal() {
...
-            return shardLimitGroupFrozen(settings);
-        } else {
-            return settings;
-        }
+    }
+
+    /**
+     * This is an illegal state, but we simulate it to capture that we do the version check
+     */
+    private Settings partial_8plusNoShardLimit() {
```
Comment (henningandersen) [inline x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgraderTests.java:137]: The 7.x backport will remove this case.

**Review by tlrx — APPROVED** [review]
LGTM

**Comment by henningandersen** [comment]
@elasticmachine update branch

**Pull request merged**